eventuals/tcp.h

```diff
@@ -0,0 +1,583 @@
+#pragma once
+
+#include "eventuals/event-loop.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace eventuals {
+namespace ip {
+namespace tcp {
+
+////////////////////////////////////////////////////////////////////////
+
+enum class Protocol {
+  NOT_SET,
+  IPV4,
+  IPV6,
+};
+
+////////////////////////////////////////////////////////////////////////
+
+class Socket {
+ public:
+  Socket(EventLoop& loop = EventLoop::Default())
+    : loop_(loop),
+      socket_(loop.io_context()) {}
+
+  Socket(const Socket& that) = delete;
+  Socket(Socket&& that)
+    : loop_(that.loop_),
+      socket_(std::move(that.socket_)),
+      protocol_(that.protocol_) {}
```

Comment:
Because you're grabbing `this` in the some of the code below you can't actually move this safely. I know we had discussed move semantics once upon a time and one of the upsides of `asio::ip::tcp::socket` is that it is moveable, but we'll have to do something else to support this. 

Ideally we could use something like `stout::borrowed_ptr` here but it doesn't have sufficient functionality yet. Instead we could add a `std::atomic<bool> borrowed_` member here which we set to true whenever we borrow `this` inside of the `Eventual` functions. Then when we're sure that we're done borrowing `this` we can set `borrowed_ = false` and then the object can be moveable again. We can `CHECK(!borrowed_.load());` in the move constructor.

Let me know if you'd like to have a synchronous discussion on this.